website/docs/configuration/resources.html.md

```diff
@@ -299,6 +300,52 @@ intended. The practice of generating multiple instances from lists should
 be used sparingly, and with due care given to what will happen if the list is
 changed later.
 
+### `for_each`: Multiple Resource Instances Defined By a Map
+
+[inpage-for_each]: #for_each-multiple-resource-instances-defined-by-a-map
+
+When the `for_each` meta-argument is present, Terraform will create instances
+based on the keys and values present in a provided map, and expose the values
+of the map to the resource for its configuration.
+
+The keys and values of the map are exposed via the `each` attribute, which can
+only be used in blocks with a `for_each` argument set.
+
+```hcl
+resource "azurerm_resource_group" "rg" {
+  for_each = {
+    a_group = "eastus"
+    another_group = "westus2"
+  }
+  name     = each.key
+  location = each.value
+}
+```
+
+The `for_each` argument also supports a set of strings; convert a list
+to a set using the `toset` function. As such, we can take the example
+in `count` and make it safer to use, as we can change items in our set
+and because the string keys are used to identify the instances,
+we will only change the items we intend to:
+
+```hcl
+variable "subnet_ids" {
+  type = list(string)
+}
+
+resource "aws_instance" "server" {
+  for_each = toset(var.subnet_ids)
+
+  ami           = "ami-a1b2c3d4"
+  instance_type = "t2.micro"
+  subnet_id     = each.key # note, each.key and each.value will be the same on a set
+
+  tags {
+    Name = "Server ${each.key}"
+  }
+}
+```
+
```

Comment:
A wise addition here would be to expand a little on what's happening to create the instances (and how they are keyed) which should help in making it clear why we don't allow `list` inputs to `for_each`, or sets of values other than `string`

Reply:
Will the actual resource names be `aws_instance.server.foo` (where foo is the key in the current iteration), or `aws_instance.server[0]`?

Reply:
Hi @wyardley, good question!

Resources created with for_each are created from a map or set, not a list, and cannot be indexed with `[0]`. They will be referenced by name, such as the `aws_instance.server.foo` example shown here.

Reply:
I've added another paragraph to this page (in this changeset) that hopefully makes this clearer!

Reply:
Correction: They'll be referenced by key using indexing syntax: `aws_instance.server["foo"]`. I'll be updating the docs to reflect this appropriately.